### ISSUE TITLE
Error obtaining redirect: invalid semicolon separator in query

### ISSUE THREAD

**Issue by timwsuqld** [oauth2-proxy/oauth2-proxy#1841]
## Expected Behavior

Redirect URL to login when authentication has expired

## Current Behavior

500 Internal Server Error

## Steps to Reproduce (for bugs)

1.  Be logged in to OTRS/Znuny, reload the page when the session has expired
2.  Get a 500 Internal Server Error, however if you remove the query parameters it works fine

## Context

The issue is due to OTRS using semicolon as a query separator. This is perfectly valid according to RFCs, even if Go doesn't support it anymore. Go needs to support it as part of the redirect URL by encoding the full URL instead of complaining.

The URL that triggers it is any OTRS url with query parameters separated by a semicolon instead of ampersand, e.g. https://support.su.org.au/otrs/index.pl?Action=AgentTicketQueue;QueueID=7;Filter=All;View=Small;SortBy=Age;OrderBy=Up;StartWindow=0;StartHit=101;TicketID=490944;TicketID=487755;TicketID=491752;TicketID=482503;TicketID=491996;TicketID=491805;TicketID=492171;TicketID=492223;TicketID=492201;TicketID=492330;TicketID=492341

Error logs:
```
oauth2_proxy-1  | [2022/10/17 23:04:25] [oauthproxy.go:904] No valid authentication in request. Initiating login.
oauth2_proxy-1  | [2022/10/17 23:04:25] [oauthproxy.go:562] Error obtaining redirect: invalid semicolon separator in query
oauth2_proxy-1  | 2022/10/17 23:04:25 http: superfluous response.WriteHeader call from github.com/oauth2-proxy/oauth2-proxy/v7/pkg/middleware.(*loggingResponse).WriteHeader (request_logger.go:77)
oauth2_proxy-1  | 2022/10/17 23:04:25 http: URL query contains semicolon, which is no longer a supported separator; parts of the query may be stripped when parsed; see golang.org/issue/25192
oauth2_proxy-1  | 10.2.37.22:58708 - b2714fd2df16960568f7876d6549b4dc - - [2022/10/17 23:04:25] support.su.org.au GET - "/otrs/index.pl?Action=AgentTicketQueue;QueueID=7;Filter=All;View=Small;SortBy=Age;OrderBy=Up;StartWindow=0;StartHit=101;TicketID=490944;TicketID=487755;TicketID=491752;TicketID=482503;TicketID=491996;TicketID=491805;TicketID=492171;TicketID=492223;TicketID=492201;TicketID=492330;TicketID=492341" HTTP/1.0 "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:106.0) Gecko/20100101 Firefox/106.0" 500 3425 0.001`

```

## Your Environment

This wasn't an issue on the `quay.io/pusher/oauth2_proxy:latest` image (5f969c90632c) from 2 years ago. The issue only started occuring when we upgraded to 7.3.0 on the `quay.io/oauth2-proxy/oauth2-proxy` image. I don't know the exact version of the old image as running with --version only gives `oauth2-proxy  (built with go1.15.5)`

- Version used: 7.3.0

**Comment by JoelSpeed** [comment]
I'm a little unsure what we are expecting here, normally a 500 wouldn't cause you to become unauthenticated, so I don't think that's what you want here. Is the problem golang or is the problem in OAuth2 Proxy?

Your application allows semi colons but golang doesn't, so what would you propose, we create a new way to encode the redirect over what we already have? Have you looked into the code to find out where exactly this happens?

**Comment by timwsuqld** [comment]
The problem is when your authentication expires, you end up with a 500 instead of being redirected correctly to the login.

A simple way to recreate this appears to be going to a URL with the semicolon separators, then in another tab signout, and then try and reload that page. Instead of the redirect to the login page, you get a 500 error page (which oddly enough is a 403 HTTP status, but showing the 500 internal error).

Clicking sign in on the error page takes you to the correct sign in page (e.g. https://support.su.org.au/oauth2/sign_in?rd=%2Fotrs%2Findex.pl%3FAction%3DAgentTicketZoom%3BTicketID%3D496971), but then attempting to signin gives you another 500 error page (with the correct 500 status this time) which I'm guessing is when it attempts to parse the redirect url. (e.g. https://support.su.org.au/oauth2/callback?state=XXXXXXXXX:/otrs/index.pl?Action%3DAgentTicketZoom;TicketID%3D496971&code=XXXXXXXXXX&scope=email%20profile%20openid%20https://www.googleapis.com/auth/userinfo.email%20https://www.googleapis.com/auth/userinfo.profile&authuser=0&hd=su.org.au&prompt=consent)

I'm not familiar with Go enough to work out where the issue is, and it appears to be 2 different code pathways triggered by the same issue (semicolons in the URL).

I'm not sure if it's Go Lang at fault, or Oauth2 Proxy, but either way we shouldn't need to be parsing the URL parameters in the redirect, we should be able to pass them through unparsed and just redirect to them at the end (unmodified). I'll try and test the older oauth2 proxy image that I believe didn't have the issue, to see if we can work out where the issue is coming from.

**Comment by timwsuqld** [comment]
I can confirm in the older image, that we didn't get a 500 error, and so the redirect to login worked. However, after the login, the redirect back to the Original URL was broken, so `/otrs/index.pl?Action=AgentTicketZoom;TicketID=168868` instead turned into `/otrs/index.pl?Action=AgentTicketZoom`.

(This is the https://quay.io/repository/pusher/oauth2_proxy?tab=tags&tag=latest latest image here, so whatever version that was).

**Comment by timwsuqld** [comment]
It sounds like this is a GoLang change, maybe something like https://pkg.go.dev/net/http#AllowQuerySemicolons will fix it? I'm not entirely sure as I've not had a chance to try and understand where in the code it's going wrong for us.

**Comment by timwsuqld** [comment]
If anyone has guidance on where in the code to make `http.AllowQuerySemicolons` work, I'm all ears. I've tried the following 2 places, and still end up with `[oauthproxy.go:616] Error obtaining redirect: invalid semicolon separator in query`. Not sure why it's not working. (go version go1.18.1)

```diff
diff --git a/oauthproxy.go b/oauthproxy.go
index 5b50bf5e..e59372b1 100644
--- a/oauthproxy.go
+++ b/oauthproxy.go
@@ -227,7 +227,7 @@ func NewOAuthProxy(opts *options.Options, validator func(string) bool) (*OAuthPr
                headersChain:       headersChain,
                preAuthChain:       preAuthChain,
                pageWriter:         pageWriter,
-               upstreamProxy:      upstreamProxy,
+               upstreamProxy:      http.AllowQuerySemicolons(upstreamProxy),
                redirectValidator:  redirectValidator,
                appDirector:        appDirector,
        }
@@ -999,7 +999,7 @@ func prepareNoCache(w http.ResponseWriter) {
 func prepareNoCacheMiddleware(next http.Handler) http.Handler {
        return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
                prepareNoCache(rw)
-               next.ServeHTTP(rw, req)
+               http.AllowQuerySemicolons(next).ServeHTTP(rw, req)
        })
 }
```

I can't find anywhere else that the http.Handler is setup, so I'm guessing there is something in the `mux` package that is doing the hard work?

**Comment by JoelSpeed** [comment]
Just out of curiosity, have you tried wrapping the entire handler? https://github.com/oauth2-proxy/oauth2-proxy/blob/8b77c9700979f2a249a9967204deecc63b91f6cf/oauthproxy.go#L265

This would mean the conversion happens as soon as the request is received, before any of the rest of our code touches the request

**Comment by timwsuqld** [comment]
@JoelSpeed thanks, I couldn't work out the right place to wrap it, and didn't realise I could wrap it there!
The following diff shows the changes that work. I guess we just need to turn this into a flag somehow so it's not a compile time change? Anyone with more experience in Go and oauth2-proxy want to tackle this?
```diff
diff --git a/oauthproxy.go b/oauthproxy.go
index 5b50bf5e..9b4b134b 100644
--- a/oauthproxy.go
+++ b/oauthproxy.go
@@ -262,7 +262,7 @@ func (p *OAuthProxy) Start() error {
 
 func (p *OAuthProxy) setupServer(opts *options.Options) error {
        serverOpts := proxyhttp.Opts{
-               Handler:           p,
+               Handler:           http.AllowQuerySemicolons(p),
                BindAddress:       opts.Server.BindAddress,
                SecureBindAddress: opts.Server.SecureBindAddress,
                TLS:               opts.Server.TLS,
```

**Comment by JoelSpeed** [comment]
I'm unlikely to have time to work on a PR any time soon, I'm still catching up from taking a bit of a break, will add help wanted and see if anyone picks it up

**Comment by timwsuqld** [comment]
This issue is still relevant, waiting for someone with more Go knowledge to turn the patch into a proper MR with options to enable/disable

**Comment by MickMake** [comment]
Hey @JoelSpeed,
I've created a pull request that fixes this issue.

See https://github.com/oauth2-proxy/oauth2-proxy/pull/2076

**Comment by timwsuqld** [comment]
This is still an issue, please reopen

**Comment by timwsuqld** [comment]
Fixed in #2248
